main/src/ca/uwaterloo/flix/language/phase/Lowering.scala

```diff
@@ -472,24 +477,192 @@ object Lowering extends Phase[Root, Root] {
+            case TypeConstructor.Int16 => SelectResult.Int16()
+            case TypeConstructor.Int32 => SelectResult.Int32()
+            case TypeConstructor.Int64 => SelectResult.Int64()
+            case TypeConstructor.BigInt => SelectResult.Reference()
+            case TypeConstructor.Str => SelectResult.Reference()
+            case TypeConstructor.RecordEmpty => SelectResult.Reference()
+            case TypeConstructor.Array => SelectResult.Reference()
+            case TypeConstructor.Channel => SelectResult.Reference() // throw InternalCompilerException(s"Channel types should already have been replaced. args: '$args''")
+            case TypeConstructor.Lazy => SelectResult.Reference()
+            case TypeConstructor.Enum(_, _) => SelectResult.Reference()
+            case TypeConstructor.Tag(_, _) => SelectResult.Reference() // throw InternalCompilerException(s"Unexpected type: '$t0'.")
+            case TypeConstructor.Native(_) => SelectResult.Reference()
+            case TypeConstructor.ScopedRef => SelectResult.Reference()
+            case TypeConstructor.Tuple(_) => SelectResult.Reference()
+            case TypeConstructor.Arrow(_) => SelectResult.Reference()
+            case TypeConstructor.RecordExtend(_) => SelectResult.Reference()
+            case TypeConstructor.Relation => SelectResult.Reference() // throw InternalCompilerException(s"Unexpected type: '$t0'.")
+            case TypeConstructor.Lattice => SelectResult.Reference() // throw InternalCompilerException(s"Unexpected type: '$t0'.")
+            case TypeConstructor.SchemaEmpty => SelectResult.Reference() // throw InternalCompilerException(s"Unexpected type: '$t0'.")
+            case TypeConstructor.SchemaExtend(_) => SelectResult.Reference() // throw InternalCompilerException(s"Unexpected type: '$t0'.")
+          }
+          case Type.Apply(_, _) => SelectResult.Reference()
+          case Type.Lambda(_, _) => SelectResult.Reference()
+          case Type.Var(_, _, _, _) => throw InternalCompilerException("Type variable (" + elementType + ") not allowed.")
+        }
+        val selectResultVariantString = selectResultVariant.toString.dropRight(2) + "Result"
+        val selectResultIdentifier = Name.Ident(sourcePositionBegin, selectResultVariantString, sourcePositionEnd)
+        val selectResultTag = Name.mkTag(selectResultIdentifier)
+        val resultSymbol = /* 2 */???
+        val selectResultPatternVariable = Pattern.Var(resultSymbol, elementType, loc)
+        val selectResultPattern = Pattern.Tag(Enums.SelectResult, selectResultTag, selectResultPatternVariable, Types.SelectResult, loc)
+        val pattern = Pattern.Tuple(List(channelIndexPattern, selectResultPattern), indexElementTupleType, loc)
+
+        // TODO: Q: Is `elementType` correct on the line below or should this be something akin to `java.lang.Object`?
+        val resultVariable = Expression.Var(resultSymbol, elementType, loc)
```

Comment:
@magnus-madsen, how does this look to you?

Reply:
I don't think this pattern matching can be quiet right/necessary. Can you explain the intuition for why you need it?